### PR TITLE
Added AbstractCorpusReader and updated JsonSerializer to optionally store explicit token character offsets

### PR DESCRIPTION
**Pull request by mssammon**
I'm working on a simple package to process a corpus with some of our tools and hand off outputs in file form for clients to integrate with outputs from other NLP tool chains. I've updated the TextAnnotationReader class to use a ResourceManager instead of explicit constructor arguments as a way of preserving its structure (constructor call to subclass-implemented initializer) while allowing more flexible representation of internal state to support the iterator functionality. 
Added a corpusreader that parses xml fragment documents, using whitespace to replace markup as a way of preserving character offsets.  (This is required by a research project collaboration). 
Added a tokenizer test that checks offset-preserving behavior of the IllinoisTokenizer.
Added a test that checks deserialization still works when new offset info is there. 
StatefulTokenizer appears to have problems with sentence endings at end of test document, but I'm leaving this issue for a separate fix: for now, the new corpusreader uses the old IllinoisTokenizer.

### CODE REVIEW

File: core-utilities/src/main/java/edu/illinois/cs/cogcomp/core/utilities/JsonSerializer.java [hunk 1]
```diff
@@ -51,6 +60,8 @@ JsonObject writeTextAnnotation(TextAnnotation ta) {
         writeString("id", ta.getId(), json);
         writeString("text", ta.getText(), json);
         writeStringArray("tokens", ta.getTokens(), json);
+        if ( doWriteTokenOffsets )
+            writeTokenOffsets( TOKENOFFSETS, ta.getView(ViewNames.TOKENS), json);
```
Comment (danyaljj) [inline core-utilities/src/main/java/edu/illinois/cs/cogcomp/core/utilities/JsonSerializer.java]: Similar behavior should be added to `readTextAnnotation`: 
https://github.com/mssammon/illinois-cogcomp-nlp-1/blob/03e49ed5aa539445579462c16410a1e2fd12ac7f/core-utilities/src/main/java/edu/illinois/cs/cogcomp/core/utilities/JsonSerializer.java#L116

File: core-utilities/src/main/java/edu/illinois/cs/cogcomp/core/utilities/JsonSerializer.java [hunk 2]
```diff
@@ -34,7 +34,16 @@
...
+    public static final String FORM = "form";
+    public static final String STARTCHAROFFSET = "startCharOffset";
+    public static final String ENDCHAROFFSET = "endCharOffset";
+    public static final String TOKENOFFSETS = "tokenOffsets";
+
+    JsonObject writeTextAnnotation(TextAnnotation ta ) {
+        return writeTextAnnotation(ta, false);
+    }
+
+    JsonObject writeTextAnnotation(TextAnnotation ta, boolean doWriteTokenOffsets ) {
```
Comment (danyaljj) [inline core-utilities/src/main/java/edu/illinois/cs/cogcomp/core/utilities/JsonSerializer.java]: Could you add tests which cover `doWriteTokenOffsets = true`? 
https://github.com/mssammon/illinois-cogcomp-nlp-1/blob/03e49ed5aa539445579462c16410a1e2fd12ac7f/core-utilities/src/test/java/edu/illinois/cs/cogcomp/annotation/TextAnnotationSerializationTest.java

File: core-utilities/src/main/java/edu/illinois/cs/cogcomp/nlp/corpusreaders/AbstractIncrementalCorpusReader.java [hunk 3]
```diff
@@ -0,0 +1,148 @@
+/**
```
Comment (danyaljj) [inline core-utilities/src/main/java/edu/illinois/cs/cogcomp/nlp/corpusreaders/AbstractIncrementalCorpusReader.java]: this is cool! Maybe we should mention it on the group email as well?

File: core-utilities/src/main/java/edu/illinois/cs/cogcomp/core/utilities/JsonSerializer.java (hunk 1 above)
Comment (mssammon) [inline core-utilities/src/main/java/edu/illinois/cs/cogcomp/core/utilities/JsonSerializer.java]: Not necessary: readTextAnnotation computes the token offsets from the raw text. The output is simply a convenience for people who want to use non-CCG code to reconstruct the annotations.

File: core-utilities/src/main/java/edu/illinois/cs/cogcomp/core/utilities/JsonSerializer.java (hunk 2 above)
Comment (mssammon) [inline core-utilities/src/main/java/edu/illinois/cs/cogcomp/core/utilities/JsonSerializer.java]: done

Comment (danyaljj) [inline core-utilities/src/main/java/edu/illinois/cs/cogcomp/core/utilities/JsonSerializer.java]: cool! Thanks!

**Comment by mssammon** [comment]
@danyaljj @christos-c this PR now includes an additional fix for WhitespaceTokenizer.  Can you review the changes and accept/request changes? -- thanks

File: corpusreaders/src/main/java/edu/illinois/cs/cogcomp/nlp/corpusreaders/ACEReader.java [hunk 4]
```diff
@@ -81,7 +81,7 @@
      * @throws Exception Exception thrown in-case of major failure.
      */
     public ACEReader(String aceCorpusHome, String[] sections, boolean is2004mode) throws Exception {
-        super(aceCorpusHome);
+        super(CorpusReaderConfigurator.buildResourceManager(aceCorpusHome));
```
Comment (christos-c) [inline corpusreaders/src/main/java/edu/illinois/cs/cogcomp/nlp/corpusreaders/ACEReader.java]: Why not pass all these arguments into the `ResourceManager`?

File: corpusreaders/src/main/java/edu/illinois/cs/cogcomp/nlp/corpusreaders/XmlFragmentWhitespacingDocumentReader.java [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+     * @param corpusName
+     * @param sourceDirectory
+     * @throws IOException
+     */
+    public XmlFragmentWhitespacingDocumentReader(String corpusName, String sourceDirectory) throws Exception {
+        super(CorpusReaderConfigurator.buildResourceManager( corpusName, sourceDirectory ));
+        taBuilder = new TokenizerTextAnnotationBuilder( new IllinoisTokenizer() );
+    }
+
+    private static ResourceManager buildResourceManager(String corpusName, String sourceDirectory) {
```
Comment (christos-c) [inline corpusreaders/src/main/java/edu/illinois/cs/cogcomp/nlp/corpusreaders/XmlFragmentWhitespacingDocumentReader.java]: Is this overriding `CorpusReaderConfigurator.buildResourceManager()`?

File: tokenizer/src/main/java/edu/illinois/cs/cogcomp/nlp/tokenizer/TokenizerStateMachine.java [hunk 6]
```diff
@@ -473,21 +473,24 @@ protected void push(State newState, int where) {
...
      */
     protected void parseText(String intext) {
-    	// find index of last non whitespace
-    	int i = intext.length()-1;
-    	for (; i >=0 ; i--) {
-    		if (!Character.isWhitespace(intext.charAt(i)))
-    			break; // found the first non-whitepsace.
-    	}
-    	i++; // length is one beyond the last whitespace char.
+    	// CHANGE (MS): preserve trailing whitespace (disabled find index of last non whitespace)
```
Comment (christos-c) [inline tokenizer/src/main/java/edu/illinois/cs/cogcomp/nlp/tokenizer/TokenizerStateMachine.java]: Is this always the desired option? Or should this be a boolean switch (`preserveWhitespace`)?

**Comment by christos-c** [comment]
@mssammon make sure the version numbers are not ported into this PR

File: tokenizer/src/main/java/edu/illinois/cs/cogcomp/nlp/tokenizer/TokenizerStateMachine.java (hunk 6 above)
Comment (mssammon) [inline tokenizer/src/main/java/edu/illinois/cs/cogcomp/nlp/tokenizer/TokenizerStateMachine.java]: I think it's always the desired option.  All this changes is the character offsets of extracted tokens; it won't by itself lead to any additional tokens being created. @cowchipkid do empty sentences ever get created? If so, does trailing whitespace decide whether or not they get created?

File: corpusreaders/src/main/java/edu/illinois/cs/cogcomp/nlp/corpusreaders/XmlFragmentWhitespacingDocumentReader.java (hunk 5 above)
Comment (mssammon) [inline corpusreaders/src/main/java/edu/illinois/cs/cogcomp/nlp/corpusreaders/XmlFragmentWhitespacingDocumentReader.java]: nope, vice versa -- this was supposed to be deleted...

File: corpusreaders/src/main/java/edu/illinois/cs/cogcomp/nlp/corpusreaders/ACEReader.java (hunk 4 above)
Comment (mssammon) [inline corpusreaders/src/main/java/edu/illinois/cs/cogcomp/nlp/corpusreaders/ACEReader.java]: future work... I may actually get to it quite soon as I'm working with an ACE Relation Extraction package right now. Presently it ignores all the usual infrastructure requirements (based on old code, and reads annotations directly from ACE source files without parsing the xml...) so integrating the standard ACEReader is a likely step.

**Comment by christos-c** [comment]
@mssammon let me know when this is ready for merging

**Comment by mssammon** [comment]
I was pretty sure I'd resolved all these conflicts... I'll take another pass...

**Comment by christos-c** [comment]
Updating version & deploying now.

**Pull request merged**